jax/_src/scipy/stats/bootstrap.py

```diff
@@ -0,0 +1,240 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import NamedTuple
+
+import scipy.stats
+import textwrap
+
+import jax
+from jax import random, vmap
+import jax.numpy as jnp
+from jax._src.scipy.special import ndtr, ndtri
+from jax._src.random import _check_prng_key
+from jax._src.numpy.util import _wraps
+
+_replace_random_state_by_key_doc = textwrap.dedent("""
+Does not support the Scipy argument ``random_state=None``.
+Instead we add an argument ``key`` which is a :class:`jax.random.PRNGKey` random key.
+"""
+)
+_replace_random_state_by_key_no_batch_jnp_statistic_doc = (
+  _replace_random_state_by_key_doc
+  + "\nDoes not support the Scipy argument ``batch``."
+    "\nIn addition, statistic is suggested to implement in JAX. Numpy implementation may lead to error."
+)
+
+
+class ConfidenceInterval(NamedTuple):
+  low: jnp.ndarray
+  high: jnp.ndarray
+
+
+class BootstrapResult(NamedTuple):
+  confidence_interval: ConfidenceInterval
+  standard_error: jnp.ndarray
+
+
+def _bootstrap_resample_and_compute_statistic(sample, statistic, n_resamples, key):
+  """
+    differences against scipy's _bootstrap_resample:
+    1. last arg `key` is jax.random.PRNGKey (v.s. int `random_state`)
+    2. `batch` arg is not needed because lax.scan auto parallelize the resampling
+    3. to save memory, it computes statistic right after resampling
+  """
+  n = sample[0].shape[-1]
+
+  @vmap
+  def _resample_and_compute_once(input_key):
+    idxs = random.randint(input_key, shape=(n,), minval=0, maxval=n)
+    # `sample` is a tuple of sample sets, we need to apply same indexing on each sample set
+    resample = jax.tree_map(lambda data: data[..., idxs], sample)
+    return statistic(*resample, axis=-1)
+
+  theta_hat_b = _resample_and_compute_once(jax.random.split(key, n_resamples))
+  # if statistics return a tuple of size n_tuple, then reshape it into (n_tuple, n_resamples)
+  #return jnp.stack(theta_hat_b)
+  return theta_hat_b.T
+
+
+def _jackknife_resample_and_compute_statistic(sample, statistic):
+  """
+    differences against scipy's _jackknife_resample:
+    1. `batch` arg is not needed because lax.scan auto parallelize the resampling
+    2. to save memory, it computes statistic right after resampling
+    """
+  n = sample.shape[-1]
+  # these assume arg `sample` has > 1 entries
+  idxs = jnp.arange(n - 1)
+  miss_first_sample = sample[1:]
+  miss_last_sample = sample[:-1]
+
+  @vmap
```

Comment:
Same comment here. Can we define `_jackknife_resample_and_compute_statistic` so it natively handles a single batch, and then use `vmap` as appropriate at the call-site?